File: python/src/store.py
"""A minimal feature store with point-in-time-correct (as-of) joins.

The hard, valuable part of any feature store is avoiding *data leakage*: when you
build a training set, each row must see only feature values that existed at or
before that row's event time -- never a value recorded afterwards. Getting this
wrong silently inflates offline metrics and wrecks the model in production.

Timestamps are integers (e.g. epoch seconds) to keep the logic portable to C#/Java.
"""

from __future__ import annotations

from bisect import bisect_right
from dataclasses import dataclass


@dataclass(frozen=True)
class FeatureValue:
    entity: str
    timestamp: int
    value: float


@dataclass(frozen=True)
class SpineRow:
    entity: str
    event_time: int


class FeatureStore:
    """Stores time-ordered feature values per entity and serves as-of joins."""

    def __init__(self) -> None:
        # entity -> (sorted timestamps, aligned values)
        self._ts: dict[str, list[int]] = {}
        self._vals: dict[str, list[float]] = {}
        self._dirty: set[str] = set()

    def ingest(self, fv: FeatureValue) -> None:
        self._ts.setdefault(fv.entity, []).append(fv.timestamp)
        self._vals.setdefault(fv.entity, []).append(fv.value)
        self._dirty.add(fv.entity)

    def _ensure_sorted(self, entity: str) -> None:
        if entity in self._dirty:
            paired = sorted(zip(self._ts[entity], self._vals[entity]))
            self._ts[entity] = [t for t, _ in paired]
            self._vals[entity] = [v for _, v in paired]
            self._dirty.discard(entity)

    def get_point_in_time(self, entity: str, event_time: int, max_staleness: int | None = None) -> float | None:
        """Latest feature value with timestamp <= event_time (as-of join, backward).

        Returns None if no such value exists, or if the newest value is older than
        `max_staleness` (feature considered expired).
        """
        if entity not in self._ts:
            return None
        self._ensure_sorted(entity)
        ts = self._ts[entity]
        idx = bisect_right(ts, event_time) - 1
        if idx < 0:
            return None
        if max_staleness is not None and event_time - ts[idx] > max_staleness:
            return None
        return self._vals[entity][idx]

    def get_training_set(self, spine: list[SpineRow], max_staleness: int | None = None) -> list[float | None]:
        """Point-in-time-correct feature column for a set of spine rows."""
        return [self.get_point_in_time(r.entity, r.event_time, max_staleness) for r in spine]

    def get_online(self, entity: str) -> float | None:
        """Latest known value for an entity (serving path)."""
        if entity not in self._ts:
            return None
        self._ensure_sorted(entity)
        return self._vals[entity][-1] if self._vals[entity] else None

File: rust/src/store.rs
//! The feature store: time-ordered values per entity plus as-of lookups.

use std::collections::{HashMap, HashSet};

/// A single observed feature value for an entity at a timestamp.
#[derive(Debug, Clone, PartialEq)]
pub struct FeatureValue {
    /// The entity the value belongs to.
    pub entity: String,
    /// When the value was observed (e.g. epoch seconds).
    pub timestamp: i64,
    /// The observed value.
    pub value: f64,
}

impl FeatureValue {
    /// Creates a feature value.
    pub fn new(entity: impl Into<String>, timestamp: i64, value: f64) -> Self {
        Self {
            entity: entity.into(),
            timestamp,
            value,
        }
    }
}

/// One row of a training spine: an entity paired with the event time whose
/// point-in-time feature value is wanted.
#[derive(Debug, Clone, PartialEq)]
pub struct SpineRow {
    /// The entity to look up.
    pub entity: String,
    /// The event time to join as-of.
    pub event_time: i64,
}

impl SpineRow {
    /// Creates a spine row.
    pub fn new(entity: impl Into<String>, event_time: i64) -> Self {
        Self {
            entity: entity.into(),
            event_time,
        }
    }
}

/// A minimal feature store that serves point-in-time-correct (as-of) joins.
#[derive(Debug, Default)]
pub struct FeatureStore {
    ts: HashMap<String, Vec<i64>>,
    vals: HashMap<String, Vec<f64>>,
    dirty: HashSet<String>,
}

impl FeatureStore {
    /// Creates an empty feature store.
    pub fn new() -> Self {
        Self::default()
    }

    /// Records a feature value. Values may arrive in any timestamp order; each
    /// entity's observations are sorted lazily on the first read.
    pub fn ingest(&mut self, fv: FeatureValue) {
        self.ts
            .entry(fv.entity.clone())
            .or_default()
            .push(fv.timestamp);
        self.vals
            .entry(fv.entity.clone())
            .or_default()
            .push(fv.value);
        self.dirty.insert(fv.entity);
    }

    /// Returns the latest feature value with `timestamp <= event_time` (a
    /// backward as-of join), or `None` if no such value exists.
    ///
    /// If `max_staleness` is `Some` and the newest eligible value is older than
    /// it, the feature is treated as expired and `None` is returned.
    pub fn get_point_in_time(
        &mut self,
        entity: &str,
        event_time: i64,
        max_staleness: Option<i64>,
    ) -> Option<f64> {
        if !self.ts.contains_key(entity) {
            return None;
        }
        self.ensure_sorted(entity);
        let ts = &self.ts[entity];
        let count = upper_bound(ts, event_time);
        if count == 0 {
            return None;
        }
        let idx = count - 1;
        if let Some(ms) = max_staleness {
            if event_time - ts[idx] > ms {
                return None;
            }
        }
        Some(self.vals[entity][idx])
    }

    /// Returns a point-in-time-correct feature column, one entry per spine row
    /// (`None` where no value is available).
    pub fn get_training_set(
        &mut self,
        spine: &[SpineRow],
        max_staleness: Option<i64>,
    ) -> Vec<Option<f64>> {
        spine
            .iter()
            .map(|r| self.get_point_in_time(&r.entity, r.event_time, max_staleness))
            .collect()
    }

    /// Returns the latest known value for an entity (the serving path), or
    /// `None` if the entity is unknown.
    pub fn get_online(&mut self, entity: &str) -> Option<f64> {
        if !self.ts.contains_key(entity) {
            return None;
        }
        self.ensure_sorted(entity);
        self.vals[entity].last().copied()
    }

    /// Orders an entity's observations by `(timestamp, value)` so that among
    /// values sharing a timestamp the largest value sorts last. This makes the
    /// as-of tie-break deterministic and independent of ingest order.
    fn ensure_sorted(&mut self, entity: &str) {
        if !self.dirty.contains(entity) {
            return;
        }
        let mut paired: Vec<(i64, f64)> = self.ts[entity]
            .iter()
            .copied()
            .zip(self.vals[entity].iter().copied())
            .collect();
        paired.sort_by(|a, b| a.0.cmp(&b.0).then_with(|| a.1.total_cmp(&b.1)));
        let new_ts: Vec<i64> = paired.iter().map(|p| p.0).collect();
        let new_vals: Vec<f64> = paired.iter().map(|p| p.1).collect();
        self.ts.insert(entity.to_string(), new_ts);
        self.vals.insert(entity.to_string(), new_vals);
        self.dirty.remove(entity);
    }
}

/// Returns the first index `i` where `ts[i] > target` (std::upper_bound).
fn upper_bound(ts: &[i64], target: i64) -> usize {
    let (mut lo, mut hi) = (0usize, ts.len());
    while lo < hi {
        let mid = (lo + hi) / 2;
        if ts[mid] <= target {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    lo
}
